rfcs/20200712-tfrt-kernel-fallback.md

```diff
@@ -0,0 +1,647 @@
+
+  // Creates a kernel with the given name and passes op_kernel_construction
+  // to kernel constructor.
+  // Returns the constructed kernel on success.
+  // In case of failure, returns a nullptr. Kernel creation can fail in one
+  // of the following cases:
+  //   1. Kernel with the given name is not found.
+  //   2. Attributes in op_kernel_construction don't match type constraints
+  //      for any of the kernels with this name.
+  //      Note that we consider a constraint to be "not matched" if attribute
+  //      it applies to is not in op_kernel_construction.
+  std::unique_ptr<TFRTOpKernel> CreateKernel(
+      StringPiece kernel_class_name,
+      TFRTOpKernelConstruction* op_kernel_construction) const;
+
+ private:
+  llvm::StringMap<std::vector<TFRTOpKernelReg>> factories_;
+};
+
+extern llvm::ManagedStatic<TFRTOpKernelFactories>
+    tfrt_forwarding_kernel_factories;
+```
+
+Similar to current TensorFlow kernel registartion, we will introduce a
+registration macro that adds a kernel to `TFRTOpKernelFactories`.
+
+```cpp
+#define REGISTER_KERNEL_FALLBACK_KERNEL(name, ...) \
+  REGISTER_KERNEL_FALLBACK_KERNEL_UNIQ_HELPER(__COUNTER__, name, __VA_ARGS__)
+
+#define REGISTER_KERNEL_FALLBACK_KERNEL_UNIQ_HELPER(ctr, name, ...) \
+  REGISTER_KERNEL_FALLBACK_KERNEL_UNIQ(ctr, name, __VA_ARGS__)
+
+#define REGISTER_KERNEL_FALLBACK_KERNEL_UNIQ(ctr, name, ...)             \
+  static bool global_tfrt_forwarding_kernel_##ctr##_registered_ = []() { \
+    ::tensorflow::tfrt_forwarding_kernel_factories->RegisterFactory(     \
+        name, TFRTOpKernelReg([](TFRTOpKernelConstruction* construction) \
+                                  -> std::unique_ptr<TFRTOpKernel> {     \
+          return std::make_unique<__VA_ARGS__>(construction);            \
+        }));                                                             \
+    return true;                                                         \
+  }();
+```
+
+## Op registration
+
+To support type specification, we will also provide a minimal Op registry and
+corresponding macro `REGISTER_KERNEL_FALLBACK_OP`.
+
+## Kernel implementation
+
+TensorFlow kernels inherit from the
+[OpKernel](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/framework/op_kernel.h;l=82?q=opkernel)
+class and depend on two key classes:
+[OpKernelConstruction](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/framework/op_kernel.h;l=256?q=opkernel)
+and
+[OpKernelContext](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/framework/op_kernel.h;l=584?q=opkernel).
+We want to provide custom implementations of these two classes in terms of data
+we get from TFRT (for e.g. inputs, attributes).
+
+There are two main approaches to customize class implementations:
+
+*   Use inheritance and define common interfaces.
+*   Use templates.
+
+We ran multiple benchmarks to get an idea of the trade offs between inheritance
```

Comment:
it's hard to follow the discussion here, since we have not described what these 2 approaches do.

one option is to defer the trade-off analysis and proposal here, till after we've described how inheritance works, and also sketched the templates based solution as well.

Reply:
I added Appendix 2 now. Does it help clarify things? (I actually looked at the two approaches in detail before sending this doc, but excluded details to reduce size of this doc)

This is the most important point of this doc. Once we decide on templates vs inheritance, we get unblocked on migrating more kernels.